src/client-side-metrics/metrics-tracer-factory.ts

```diff
@@ -0,0 +1,600 @@
+} = require('@opentelemetry/sdk-metrics');
+import * as Resources from '@opentelemetry/resources';
+import {MetricExporter} from '@google-cloud/opentelemetry-cloud-monitoring-exporter';
+import * as ResourceUtil from '@google-cloud/opentelemetry-resource-util';
+import {ObservabilityOptions} from './observability-options';
+import * as fs from 'fs';
+
+/**
+ * Information about a Bigtable operation.
+ */
+interface OperationInfo {
+  /**
+   * The number of retries attempted for the operation.
+   */
+  retries?: number;
+  /**
+   * The final status of the operation (e.g., 'OK', 'ERROR').
+   */
+  finalOperationStatus: string;
+  /**
+   * Number of times a connectivity error occurred during the operation.
+   */
+  connectivityErrorCount?: number;
+  streamingOperation: string;
+}
+
+/**
+ * Information about a single attempt of a Bigtable operation.
+ */
+interface AttemptInfo {
+  /**
+   * The final status of the attempt (e.g., 'OK', 'ERROR').
+   */
+  finalOperationStatus: string;
+  /**
+   * Whether the operation is a streaming operation or not
+   */
+  streamingOperation: string;
+}
+
+/**
+ * A collection of OpenTelemetry metric instruments used to record
+ * Bigtable client-side metrics.
+ */
+interface Metrics {
+  operationLatencies: typeof Histogram;
+  attemptLatencies: typeof Histogram;
+  retryCount: typeof Counter;
+  applicationBlockingLatencies: typeof Histogram;
+  firstResponseLatencies: typeof Histogram;
+  serverLatencies: typeof Histogram;
+  connectivityErrorCount: typeof Histogram;
+  clientBlockingLatencies: typeof Histogram;
+}
+
+/**
+ * An interface representing a Date-like object.  Provides a `getTime` method
+ * for retrieving the time value in milliseconds.  Used for abstracting time
+ * in tests.
+ */
+interface DateLike {
+  /**
+   * Returns the time value in milliseconds.
+   * @returns The time value in milliseconds.
+   */
+  getTime(): number;
+}
+
+/**
+ * Interface for a provider that returns DateLike objects. Used for mocking dates in tests.
+ */
+interface DateProvider {
+  /**
+   * Returns a DateLike object.
+   * @returns A DateLike object representing the current time or a fake time value.
+   */
+  getDate(): DateLike;
+}
+
+/**
+ * The default DateProvider implementation.  Returns the current date and time.
+ */
+class DefaultDateProvider {
+  /**
+   * Returns a new Date object representing the current time.
+   * @returns {Date} The current date and time.
+   */
+  getDate() {
+    return new Date();
+  }
+}
+
+/**
+ * An interface representing a tabular API surface, such as a Bigtable table.
+ */
+export interface ITabularApiSurface {
+  instance: {
+    id: string;
+  };
+  id: string;
+  bigtable: {
+    appProfileId?: string;
+  };
+}
+
+const packageJSON = fs.readFileSync('package.json');
+const version = JSON.parse(packageJSON.toString()).version;
+
+/**
+ * A class for tracing and recording client-side metrics related to Bigtable operations.
+ */
+class MetricsTracer {
```

Comment:
It looks like this class is tightly coupled to OpenTelemetry metrics. I really think you should add a layer of abstraction, so we can change things out easier in the future

If you do keep it coupled, maybe rename this to OpenTelemetryMetricsTracer, so it's more clear what this class is doing